python/cvmfs/repository.py

```diff
@@ -197,39 +196,230 @@ def _pop_catalog(self):
+
+    def add(self, file_name):
+        full_path = os.path.join(self._cache_dir, file_name)
+        tmp_dir = self.get_transaction_dir()
+        return Resource(full_path, "w+", tmp_dir)
+
+    def get(self, file_name):
+        full_path = os.path.join(self._cache_dir, file_name)
+        if os.path.exists(full_path):
+            return open(full_path, 'rb')
+        return None
+
+    def clear(self):
+        data_path = os.path.join(self._cache_dir, 'data')
+        shutil.rmtree(data_path)
+        self._create_cache_structure()
+
+
+class Fetcher(object):
+    """ Abstract wrapper around a Fetcher """
+
+    __metadata__ = abc.ABCMeta
+
+    def __init__(self, source, cache_dir=''):
+        self.__cache = Cache(cache_dir)
+        self.source = source
+
+    def _make_file_uri(self, file_name):
+        return os.path.join(self.source, file_name)
+
+    def get_cache_path(self):
+        return self.__cache.get_cache_path()
+
+    def retrieve_file(self, file_name):
+        """ Method to retrieve a file from the cache if exists, or from
+        the repository if it doesn't
+        """
+        cached_file = self.__cache.get(file_name)
+        if not cached_file:
+            cached_file = self.__cache.add(file_name)
+            self._retrieve_file(file_name, cached_file)
+            return self.__cache.get(file_name)
+        return cached_file
+
+    def retrieve_raw_file(self, file_name):
+        cached_file = self.__cache.get(file_name)
+        if not cached_file:
+            cached_file = self.__cache.add(file_name)
+            self._retrieve_raw_file(file_name, cached_file)
+            return self.__cache.get(file_name)
+        return cached_file
+
+    @abc.abstractmethod
+    def _retrieve_file(self, file_name, cached_file):
+        """ Abstract method to retrieve a file from the repository """
+        pass
+
+    @abc.abstractmethod
+    def _retrieve_raw_file(self, file_name, cached_file):
+        """ Abstract method to retrieve a raw file from the repository """
+        pass
+
+
+class LocalFetcher(Fetcher):
+    """ Retrieves files only from the local cache """
+
+    def __init__(self, local_repo, cache_dir=''):
+        super(LocalFetcher, self).__init__(local_repo, cache_dir)
+
+    def _retrieve_file(self, file_name, cached_file):
+        full_path = self._make_file_uri(file_name)
+        if os.path.exists(full_path):
+            compressed_file = open(full_path, 'r')
+            decompressed_content = zlib.decompress(compressed_file.read())
+            compressed_file.close()
+            cached_file.write(decompressed_content)
+            cached_file.close()
+        else:
+            raise FileNotFoundInRepository(file_name)
+
+    def _retrieve_raw_file(self, file_name, cached_file):
+        """ Retrieves the file directly from the source """
+        full_path = self._make_file_uri(file_name)
+        if os.path.exists(full_path):
+            compressed_file = open(full_path, 'r')
```

Comment:
Retrieving raw files doesn't necessarily mean that they are compressed, no? But the variable name suggests it.